routing/http/server/server.go

```diff
@@ -236,13 +241,78 @@ func (s *server) findProvidersJSON(w http.ResponseWriter, provIter iter.ResultIt
 		return
 	}
 
+	filteredProviders := filterProviders(providers, filterAddrs, filterProtocols)
+
 	writeJSONResult(w, "FindProviders", jsontypes.ProvidersResponse{
-		Providers: providers,
+		Providers: filteredProviders,
 	})
 }
+func (s *server) findProvidersNDJSON(w http.ResponseWriter, provIter iter.ResultIter[types.Record], filterAddrs, filterProtocols []string) {
```

Comment:
@2color Any reason why we duplicate this code other than just having filtering on /providers and not /peers? 

Thoughts on adding filtering support to /peers and not having this duplicated handler? It feels useful for peer lookups too, at least filter-addrs one ("give me only /wss and /webrtc of this peer")

Reply:
Yeah that's a good idea. One challenge was doing so with generics so I just focused on making it work first. 

I'll give it another look.

Reply:
I agree we should enable filtering to the `/peers`.

I had another look at this and haven't been able to find a good solution to reuse the code, due to the combination of generics and the different return types (types.Record and types.PeerRecord)

The main open question I have is why `FindProviders` and `FindPeers` have a different return type if they both return a list of `PeerRecords`:

```
FindProviders(ctx context.Context, cid cid.Cid, limit int) (iter.ResultIter[types.Record], error)
FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[*types.PeerRecord], error)
``` 

Do you have an idea @lidel?

Reply:
- `/routing/v1/peers` will always return results that use Peer schema.
- `/routing/v1/providers` may return different schemas, [Peer schema is not the only one](https://specs.ipfs.tech/routing/http-routing-v1/#known-schemas),  so a more generic type is used to allow pass-through in places like https://delegated-ipfs.dev/ which proxy results from other systems.
  - fysa `cid.contact/routing/v1` switched to Peer schema some time ago, making it now confusing why we dont use Peer schema everywhere, and why we waste space for Schema field in every response, but this is [probably both true] (a) a way of having open-ended API that can be evolved by community without breaking existing clients (b) reality of different teams working on different things with different priorities. 

Personally I see value in (a), I think we should keep `FindProviders(ctx context.Context, cid cid.Cid, limit int) (iter.ResultIter[types.Record], error)` return generic `Record`, to ensure   `Schema` field is not waste of space, to allow others to experiment without having to invent yet another API.

Reply:
@2color  To address your task at hand, something we could do:
- keep interface as-is, write reusable things so they return Record, and in FindPeers have a wrapper that translates Records to PeerRecord for now (this way we dont break API)
- make a breaking change and document in release notes: make `FindPeers` return generic `Record` as a way of deduplicating code boilerplate. Unlikely we will see different schema on `/peers`, but would not hurt to have that ability.

Reply:
I addressed your feedback taking the first approach and avoiding breaking changes. It felt a bit messy, but we haven't broken any public interfaces.